lib/data_stores/mysql_advanced_data_store.py

```diff
@@ -0,0 +1,636 @@
+        result[subject] = values
+        if limit:
+          limit -= len(values)
+
+      if limit and limit <= 0:
+        break
+
+    return result.iteritems()
+
+  def MultiSet(self, subject, values, timestamp=None, token=None, replace=True,
+               sync=True, to_delete=None):
+    """Set multiple predicates' values for this subject in one operation."""
+    self.security_manager.CheckDataStoreAccess(token, [subject], "w")
+    to_delete = set(to_delete or [])
+    if timestamp is None:
+      timestamp = time.time() * 1e6
+
+    # Prepare a bulk insert operation.
+    subject = utils.SmartUnicode(subject)
+    to_insert = []
+    to_replace = []
+
+    # Build a document for each unique timestamp.
+    for attribute, sequence in values.items():
+      for value in sequence:
+        entry_timestamp = None
+
+        if isinstance(value, tuple):
+          value, entry_timestamp = value
+
+        if entry_timestamp is None:
+          entry_timestamp = timestamp
+
+        predicate = utils.SmartUnicode(attribute)
+        data = self._EncodeValue(value)
+
+        # Replacing means to delete all versions of the attribute first.
+        if replace or attribute in to_delete:
+          duplicates = self._CountDuplicateAttributes(subject, predicate)
+          if duplicates > 1:
+            to_delete.add(attribute)
+            to_insert.extend(
+                [subject, predicate, data, int(entry_timestamp)])
+          else:
+            if attribute in to_delete:
+              to_delete.remove(attribute)
+            if duplicates == 0:
+              to_insert.extend(
+                  [subject, predicate, data, int(entry_timestamp)])
+            elif duplicates == 1:
+              to_replace.extend(
+                  [subject, predicate, data, int(entry_timestamp)])
+        else:
+          to_insert.extend(
+              [subject, predicate, data, int(entry_timestamp)])
+
+    if to_delete:
+      self.DeleteAttributes(subject, to_delete, token=token)
+
+    if to_replace:
+      transaction = self._BuildReplaces(to_replace)
+      self._ExecuteInsert(transaction)
+
+    if to_insert:
+      if sync:
+        transaction = self._BuildInserts(to_insert)
+        self._ExecuteInsert(transaction)
+      else:
+        with self.lock:
+          self.to_insert.extend(to_insert)
+
+  def Flush(self):
+    with self.lock:
+      to_insert = self.to_insert
+      self.to_insert = []
+
+    if to_insert:
+      transaction = self._BuildInserts(to_insert)
+      self._ExecuteInsert(transaction)
+
+  def _CountDuplicateAttributes(self, subject, predicate):
+    query = "SELECT count(*) AS total FROM aff4 WHERE \
+subject_hash=unhex(md5(%s)) AND attribute_hash=unhex(md5(%s))"
+    args = [subject, predicate]
+    result = self._ExecuteQuery(query, args)
+    return int(result[0]["total"])
+
+  def _BuildReplaces(self, values):
+    transaction = []
+    value_sets = zip(values[0::4], values[1::4], values[2::4], values[3::4])
```

Comment:
I will look into this.  Looking at it now I'm not sure why I didn't just make a list of tuples in the first place.  I think at the time there was something preventing me from altering the MultiSet behavior for building those lists, but I'm not seeing it now.